### ISSUE TITLE
Receiving ObjectStoreFullError when running Brian2 simulations with Mopet

### ISSUE THREAD

**Issue by lu-kolb** [caglorithm/mopet#10]
The error is thrown when ray's ObjectStore is full. This problem is described in [FAQ of ray](https://docs.ray.io/en/master/memory-management.html#frequently-asked-questions-faq).

Proposed solution is to enable [LRU fallback](https://docs.ray.io/en/master/memory-management.html#enabling-lru-fallback):
```python
ray.init(lru_evict=True)
```

LRU strategy was used in previous versions of ray, FAQ suggests to enable it again for applications that keep references to all objects ever created (such as iPython/JupyterNotebook).

As this library is focused mainly on scientific computing where many use Jupyter Nootebok, I would suggest to set enable `lru_evict=True` per default and allow in the `Exploration.__init__()` to disable it.

<img width="1473" alt="Bildschirmfoto 2020-06-05 um 10 18 24" src="https://user-images.githubusercontent.com/7281685/83859042-aa396600-a71d-11ea-82df-de38e936059b.png">

**Comment by caglorithm** [comment]
Thank you very much for the issue! I have been running ray in Jupyter Lab form the beginning but haven't encountered this problem yet. The ray FAQ that you've linked says about `lru_evict=True`:
> Please note that relying on this is not recommended - instead, if possible you should try to remove references as they’re no longer needed in your application to free space in the object store.

What kind of scenario causes this problem for you?

Referring to the quote above, maybe a better solution could be to add an option to enable it when necessary. What do you think?

**Comment by lu-kolb** [comment]
I'm running Brian2 simulations with mopet on my local machine, I only have this issue once I run large simulations that need more than 1 hour to complete and consist of many runs (in my case ~900). I did not have any problems with small simulations (~30-100 runs) either.

Oh I missed the part you quoted, yes sounds like a good idea to make it optional when it's not recommended. I would suggest to add it to the README for iPython and Jupyter Notebook users, so they are aware of what to do when confronted with the error.

Setting `lru_evict=True` resolved this issue for me. I will try to reproduce the error outside of iPython and Jupyter Notebook to assure that they are the actual culprit. :)

**Comment by lu-kolb** [comment]
I can reproduce the behavior now by using the normal python interpreter, so no iPython or Jupyter Notebook. It is not yet clear why this happens though. But it's not exclusively due to iPython or Jupyter Notebook.

It occurs only after many simulations have been run. Assuming that ray cleans up each python job completely and with it all the memory used by the run function, it might occur because of result dicts that stay in memory until mopet writes them to the hdf5 file. But that is just a broad guess. Have to dive into it first. ;)

**Comment by caglorithm** [comment]
It would be nice if we can find a simple scenario in which the error occurs... maybe an example that creates a billion ray jobs?

**Comment by lu-kolb** [comment]
## Minimal Example
```python
from mopet import mopet

def run(params):
    import time
    x = 2
    t_end = time.time() + 5
    while time.time() < t_end:
        y = x * x

    return {"my_list": [1, 2, 3]}


def test():
    params = {"a": np.arange(0, 100, 1)}

    ex = mopet.Exploration(run, params, exploration_name="testC")
    ex.run(num_cpus=1)
    ex.load_results()

```

If you run the following command you can observe how the object store still stores all job results even after they have been written to the hdf5 file. In the screenshot 6 jobs completed (-> 27) and the rest is still in progress (-> ?).
```shell
watch -n 1 ray memory
```

In the example I limited the CPUS to 1, so only 1 job should run. Each job runs for 5 seconds, gets written to hdf5 file and should then ideally be removed from ray's object store.
But, the `ray_returns` still holds for each run an ObjectId reference. However, [they must be removed once we don't need them](https://docs.ray.io/en/latest/memory-management.html#frequently-asked-questions-faq) which would be after writing the results to disk.

<img width="1074" alt="Bildschirmfoto 2020-06-19 um 16 16 34" src="https://user-images.githubusercontent.com/7281685/85142187-44ce9480-b248-11ea-9b43-bf02782e985b.png">


My solution to this is:
```python
for run_id, ray_return in tqdm.tqdm(ray_returns.items()):
     ...
    # removes ObjectId reference and ray can evict results of job from object store.
     ray_returns[run_id] = None
```

**Comment by caglorithm** [comment]
Congrats! This is super simple and makes sense. We should merge this!

**Comment by caglorithm** [comment]
@lrebscher, thought this was merged already?

**Comment by lu-kolb** [comment]
yes it is merged, was not sure if issue should stay open until it has been released. But it makes more sense to close it directly.

**Comment by caglorithm** [comment]
We could do a new release update, true! Do you have any PR in the pipeline or should I just do it?

**Comment by lu-kolb** [comment]
@caglorithm I would finish branch `13_exception_handling` in a couple of days, but otherwise I have no immediate PRs incoming. I leave it up to you!
